Revise 3rd party integrations for new implementations

kb2ma opened the pull request:
I have revised the 3rd party integrations pages based on new cloud IoT provisioning implementations, and also a balenaBlock (Cloud Relay) to use the provisioning and send data. Below is some background on the provisioning implementations, and how that changed the structure of the integration docs.

## Provisioning implementation

The revised cloud IoT integrations are in these repositories: [aws-iot-provision](https://github.com/balena-io-examples/aws-iot-provision), [azure-iot-provision](https://github.com/balena-io-examples/azure-iot-provision), and [gcp-iot-provision](https://github.com/balena-io-examples/gcp-iot-provision). They all now use the same approach to provisioning a device:

1. HTTP request to cloud
1. Run cloud function, which executes the steps below
1. Validate UUID
1. Generate public key pair / certificate
1. Register with IoT Core
1. Push credentials as env vars to balenaCloud

Cloud Relay block makes it easy to use the provisioning tools and send data. It automatically calls the HTTP endpoint on the cloud to provision the device. The block then expects data from a local MQTT broker, and relays it to the provisioned IoT service in the cloud.

## Integration pages

Since the cloud provisioning implementations are similar and they provide detailed setup instructions, this PR condenses the documentation to a single dynamic page. The page describes how provisioning works and also mentions Cloud Relay block. It then directs the user to the implementation repository for generic cloud setup documentation and specific provisioning setup documentation.

This PR also includes a second commit to remove the original 3rd party integration pages.

kb2ma commented:
Added commit with the link to the blog post, so removed draft label.

kb2ma commented:
@vipulgupta2048 , I'd appreciate your review at least from the perspective of how to implement a dynamic page.

vipulgupta2048 commented on config/dictionaries/cloud.json:
```diff
@@ -0,0 +1,56 @@
+    "cloudFunctionName": "Lambda function",
+    "cloudFunctionFullNameStrong": "**AWS Lambda (cloud) function**",
+    "accountSetupLink": "https://github.com/balena-io-examples/aws-iot-provision#aws-setup",
+    "toolsSetupLink": "https://github.com/balena-io-examples/aws-iot-provision#tools-setup"
```
```suggestion
    "toolsSetupLink": "https://github.com/balena-io-examples/aws-iot-provision#development-setup"
```

If you modify this heading to my suggestion then you wouldn't need these properties

kb2ma replied:
I was trying to make a distinction between generic platform setup and tools setup. Let's discuss.

kb2ma replied:
We decided to leave as is.

vipulgupta2048 commented on config/dictionaries/cloud.json:
```diff
@@ -0,0 +1,56 @@
+    "iotCoreName": "IoT Core",
+    "iotCoreUrl": "https://cloud.google.com/iot-core",
+    "credentialType": "public key pair",
+    "provisionRepoName": "gcp-iot-provision",
+    "provisionRepoUrl": "https://github.com/balena-io-examples/gcp-iot-provision",
```
Same for these properties, I think we can optimize the properties further to reduce static data in this dictionary.

kb2ma replied:
Reduced dictionary size from 16 to 9 in 616c00c.

vipulgupta2048 commented on config/redirects.txt:
```diff
@@ -109,11 +108,9 @@
 ^\/reference\/OS\/overview\/2.x\/?(\?.+)?(#.+)?$ -> /reference/OS/overview/$1$2
 ^\/reference\/OS\/network\/1.x\/?(\?.+)?(#.+)?$ -> /reference/OS/network/$1$2
 ^\/reference\/OS\/network\/2.x\/?(\?.+)?(#.+)?$ -> /reference/OS/network/$1$2
-^\/learn\/manage\/app-types\/?(\?.+)?(#.+)?$ -> /learn/manage/fleet-types/$1$2
```
Again, not sure why we are deleting redirects

kb2ma replied:
More sloppy conflict resolution.

vipulgupta2048 commented on pages/learn/develop/integrations/cloud-iot-provisioning.md:
```diff
@@ -0,0 +1,45 @@
+  variables: [ $cloud ]
+  ref: $original_ref/$cloud
+  $switch_text: Cloud IoT Provisioning for $cloud
+---
```
```suggestion
---

```

kb2ma replied:
Fixed in 771cf24.

vipulgupta2048 commented on pages/learn/develop/integrations/cloud-iot-provisioning.md:
```diff
@@ -0,0 +1,45 @@
+dynamic:
+  variables: [ $cloud ]
+  ref: $original_ref/$cloud
+  $switch_text: Cloud IoT Provisioning for $cloud
```
```suggestion
  $switch_text: I want to provision Cloud IoT device for $cloud
```

kb2ma replied:
Good idea. I see how your wording is parallel to that in the Get Started section for a device.

kb2ma replied:
Fixed in 771cf24.

vipulgupta2048 reviewed (commented):
Hey @kb2ma Good PR, I reviewed the configuration. Might need a deeper review of the content, maybe we can sync on this.

kb2ma commented on config/redirects.txt:
```diff
@@ -109,11 +108,9 @@
 ^\/deployment\/build-optimisation\/?(\?.+)?(#.+)?$ -> /learn/deploy/build-optimization/$1$2
 ^\/reference\/data-api\/?(\?.+)?(#.+)?$ -> /reference/api/overview/$1$2
 ^\/reference\/resinOS\/?(\?.+)?(#.+)?$ -> /reference/OS/$1
-^\/reference\/OS\/meta-balena\/?(\?.+)?(#.+)?$ -> /reference/OS/customer-board-support/$1$2
-^\/reference\/OS\/custom-device-support\/?(\?.+)?(#.+)?$ -> /reference/OS/customer-board-support/$1$2
```
This removal was sloppy conflict resolution when rebasing locally. Thanks for catching this error.

kb2ma commented on pages/learn/develop/integrations/cloud-iot-provisioning.md:
```diff
@@ -0,0 +1,45 @@
+---
+# Cloud IoT Provisioning for {{ $cloud.name }}
+
+The [{{ $cloud.fullPlatformName }}]({{ $cloud.platformUrl }}) platform provides a valuable suite of services to collect, store, and distribute IoT data and actions. Its [{{ $cloud.iotCoreName }}]({{ $cloud.iotCoreUrl }}) service is the portal for registration and messaging with Internet-connected things. We want to make it easy for balena devices to register and interact with {{ $cloud.iotCoreNam ...
```
Not a big deal, but let's discuss.

kb2ma replied:
Left as is.

kb2ma commented:
Thanks for the review @vipulgupta2048 ! I fixed the straightforward issues, but there are a couple worth discussing. I'll set up a meeting and loop in @mpous as well.

vipulgupta2048 commented:
Thanks @kb2ma Will draft this so it doesn't get merged before the discussion

kb2ma commented:
@vipulgupta2048 integrated the changes as discussed recently. Dictionary size per cloud reduced from 16 items to 9.  I also updated the text to accommodate the dictionary changes, and made a few other minor improvements.

Also, thanks for mentioning the continued presence of obsolete files. I confirmed this and removed in 69c7533.

Checks passed, so ready for hopefully final review.

vipulgupta2048 commented on config/redirects.txt:
```diff
@@ -52,7 +52,6 @@
 ^\/using\/cli\/?(\?.+)?(#.+)?$ -> /reference/balena-cli/$1$2
 ^\/learn\/deploy\/release-strategy\/update-locking\/API\/?(\?.+)?(#.+)?$ -> /reference/supervisor/supervisor-api/$1$2
 ^\/reference\/cli\/?(\?.+)?(#.+)?$ -> /reference/balena-cli/$1$2
-^\/learn\/develop\/integrations\/google\/?(\?.+)?(#.+)?$ -> /learn/develop/integrations/google-iot/$1$2
```
@kb2ma The previous integrations were quite old, but I think it would make sense for them to be redirected to the new page? Can you add 4 redirect rules to redirect the 4 integrations over to the new page.

kb2ma replied:
Fixed in 83b1b5b

vipulgupta2048 commented on config/redirects.txt:
```diff
@@ -116,4 +115,4 @@
 ^\/reference\/OS\/network\/2.x\/?(\?.+)?(#.+)?$ -> /reference/OS/network/$1$2
-^\/learn\/manage\/app-types\/?(\?.+)?(#.+)?$ -> /learn/manage/fleet-types/$1$2
\ No newline at end of file
+^\/learn\/manage\/app-types\/?(\?.+)?(#.+)?$ -> /learn/manage/fleet-types/$1$2
```
Not sure what change was made to this line?

kb2ma replied:
The change was whitespace added at the end of the line. Fixed in 6d8c525.

vipulgupta2048 commented on pages/learn/develop/integrations.md:
```diff
@@ -5,11 +5,11 @@ excerpt: Guides to integrate {{ $names.company.lower }} with a variety of IoT pl
 
-### Guides
+### Tools
+
```
Why redirect by default to AWS?
Why mention lambda specifically? Are you trying to provide customers with a reference or we lean more closely to AWS?

kb2ma replied:
No longer applicable.

vipulgupta2048 commented on pages/learn/develop/integrations.md:
```diff
@@ -5,11 +5,11 @@ excerpt: Guides to integrate {{ $names.company.lower }} with a variety of IoT pl
 
 # Integrate with 3rd parties
```
@kb2ma Why do we even have this page, and then the `cloud-iot-provisioning.md` file in a directory. I think this document doesn't provide no value? Rather this intro could be added to the other file instead

kb2ma replied:
Thanks for the fresh eyes on this. I agree there isn't sufficient purpose for a separate 3rd party integrations page. Created a "Provision with cloud IoT" navigation entry, which links directly to `cloud-iot-provisioning.md`. Also removed the `integrations` directory. See fd5f464.

vipulgupta2048 commented on config/navigation.txt:
```diff
@@ -48,10 +48,7 @@ Learn
       [/learn/develop/hardware/usb]
       [/learn/develop/hardware/i2c-and-spi]
     [/learn/develop/integrations]
```
If you wish to execute my suggestion to remove this page, then you can remove this page from here as well..

kb2ma replied:
Fixed in fd5f464.

vipulgupta2048 commented on pages/learn/develop/integrations.md:
```diff
@@ -5,11 +5,11 @@ excerpt: Guides to integrate {{ $names.company.lower }} with a variety of IoT pl
+
+[Cloud IoT Provisioning](cloud-iot-provisioning/aws) -- A cloud function (lambda) to generate a public key identity for a balena device in a cloud's IoT registry. Stores the key in the registry and on the device. Works with AWS, Microsoft Azure, and Google Cloud.
+
+<a href="https://github.com/balena-io-examples/cloud-relay#readme" target="_blank">Cloud Relay block</a> -- Triggers the provisioning tool above from a balena device to bootstrap identity, and then securely sends MQTT data to the cloud.
```
```suggestion
The <a href="https://github.com/balena-io-examples/cloud-relay#readme" target="_blank">Cloud Relay block</a> -- Triggers the provisioning tool above from a balena device to bootstrap identity, and then securely sends MQTT data to the cloud.
```

kb2ma replied:
No longer applicable.

vipulgupta2048 reviewed (commented):
Looking really good @kb2ma Found some more improvements we could and definitely a page to remove. Do let me know what you think? When the changes get resolved, I think this should be ready to go.

vipulgupta2048 commented:
@balena-ci rebase

kb2ma commented:
@vipulgupta2048 thanks for the improvements. See inline comments. I did remove the separate 3rd party integrations page.

vipulgupta2048 commented:
@kb2ma Thanks, please do make a note to resolve the comments when you are finished with them. You can only leave the ones pending discussions or a question you have. Much cleaner to re-review.

vipulgupta2048 reviewed (approved):
Great work here @kb2ma Thanks for all the review implementations. Please squash the commits before removing the draft status on this PR.

kb2ma commented:
Thanks @vipulgupta2048  for the tip on resolving comments for re-review.
